### PR TITLE
Implements the attention kernel with vertical and slash sparse pattern described in Appendix C.4.2 of https://arxiv.org/abs/2407.02490 (as sparse_attn_func)

### PR DESCRIPTION
**Pull request by minminsun**
Implements the attention kernel with vertical and slash sparse pattern described in Appendix C.4.2 of https://arxiv.org/abs/2407.02490 (as sparse_attn_func).

### CODE REVIEW

File: tests/flash_attn_wrapper.py [hunk 1]
```diff
@@ -101,3 +102,33 @@ def _(
         softcap: float = 0.0,
 ) -> torch.Tensor:
     return torch.empty_like(decode_query)
+
+@torch.library.custom_op("vllm::sparse_attn_func", mutates_args=[])
```
Comment (LucasWilkinson) [inline tests/flash_attn_wrapper.py:106]: If we are not testing `opcheck` (you'll see `torch.library.opcheck` in some of the other tests) im not sure there's much value in registering this as a custom-op here. Im not exactly sure what the `opcheck`s are doing here given its testing the custom op wrapper that only exists in the test code. @ProExpertProg do you have any context here?

File: vllm_flash_attn/flash_attn_interface.py [hunk 2]
```diff
@@ -112,6 +136,46 @@ def _flash_attn_varlen_forward(
     )
     return out, softmax_lse
 
+def sparse_attn_func(
```
Comment (LucasWilkinson) [inline vllm_flash_attn/flash_attn_interface.py:139]: Can you add a doc string like `flash_attn_func` here, I think it would be a helpful record about what the new arguments do (namely `block_count`, `block_offset`, `column_count`, `column_index`)

File: csrc/flash_attn/flash_api.cpp [hunk 3]
```diff
@@ -261,6 +336,203 @@ void set_params_alibi(Flash_fwd_params &params, const c10::optional<at::Tensor>
...
+        out = out.index({"...", torch::indexing::Slice(torch::indexing::None, head_size_og)});
+        if (out_.has_value()) { out_.value().copy_(out); }
+    }
+
+    if (seqlenq_ngroups_swapped) {
+        out = out.transpose(1, 2).reshape({batch_size, 1, num_heads_k * seqlen_q, head_size_og});
+        // NOTE(woosuk): The two lines are not needed because out_padded and q_padded are not used.
+        // out_padded = out_padded.transpose(1, 2).reshape({batch_size, 1, num_heads_k * seqlen_q, head_size_og});
+        // q_padded = q_padded.transpose(1, 2).reshape({batch_size, 1, num_heads_k * seqlen_q, head_size_og});
+        softmax_lse = softmax_lse.reshape({batch_size, num_heads_k * seqlen_q, 1});
```
Comment (LucasWilkinson) [inline csrc/flash_attn/flash_api.cpp:531]: this may need: https://github.com/vllm-project/flash-attention/pull/34

File: csrc/flash_attn/flash_api.cpp [hunk 4]
```diff
@@ -154,6 +154,70 @@ void set_params_fprop(Flash_fwd_params &params,
...
+        window_size_left,
+        window_size_right,
+        softcap,
+        seqlenq_ngroups_swapped,
+        unpadded_lse
+    );
+    params.block_count = (int*)block_count.data_ptr();
+    params.block_offset = (int*)block_offset.data_ptr();
+    params.column_count = (int*)column_count.data_ptr();
+    params.column_index = (int*)column_index.data_ptr();
```
Comment (LucasWilkinson) [inline csrc/flash_attn/flash_api.cpp:212]: nit: can you use `.const_data_ptr<int>()` here for some basic type checking?

**Review by LucasWilkinson — COMMENTED** [review]
vllm-flash-attn has unfortunately diverged pretty drastically from the upstream but we are trying to simplify diffs with upstream and in that spirit I think it would be really helpful if we could push most of the additions in
`csrc/flash_attn/src/flash_fwd_kernel.h` and `csrc/flash_attn/src/flash_fwd_launch_template.h` 
into their own file, i.e. move them to files like:
`csrc/flash_attn/src/vllm_extensions/flash_fwd_sparse_kernel.h` and `csrc/flash_attn/src/vllm_extensions/flash_fwd_sparse_launch_template.h`

(Im looking into ways to reduce diffs in `csrc/flash_attn/flash_api.cpp` but this is trickier so I think what's in this PR currently is fine, we can address it in a future PR)

**Comment by minminsun** [comment]
> vllm-flash-attn has unfortunately diverged (in conflicting ways) from the upstream but we are trying to simplify diffs with upstream and in that spirit I think it would be really helpful if we could push most of the additions in `csrc/flash_attn/src/flash_fwd_kernel.h` and `csrc/flash_attn/src/flash_fwd_launch_template.h` into their own file, i.e. move them to files like: `csrc/flash_attn/src/vllm_extensions/flash_fwd_sparse_kernel.h` and `csrc/flash_attn/src/vllm_extensions/flash_fwd_sparse_launch_template.h`
> 
> (Im looking into ways to reduce diffs in `csrc/flash_attn/flash_api.cpp` but this is trickier so I think what's in this PR currently is fine, we can address it in a future PR)

Thanks for your suggestion! I've moved code to new files.

File: tests/flash_attn_wrapper.py (hunk 1 above)
Comment (minminsun) [reply tests/flash_attn_wrapper.py]: OK, the custom-op registering has been removed.

File: vllm_flash_attn/flash_attn_interface.py (hunk 2 above)
Comment (minminsun) [reply vllm_flash_attn/flash_attn_interface.py:139]: Ok, done.

File: csrc/flash_attn/flash_api.cpp (hunk 3 above)
Comment (minminsun) [reply csrc/flash_attn/flash_api.cpp:531]: Rebased.

File: csrc/flash_attn/flash_api.cpp (hunk 4 above)
Comment (minminsun) [reply csrc/flash_attn/flash_api.cpp]: Thanks for the suggestion. Done.

File: vllm_flash_attn/flash_attn_interface.py [hunk 5]
```diff
@@ -112,6 +136,78 @@ def _flash_attn_varlen_forward(
...
+        v: (batch_size, seqlen, nheads_k, headdim)
+        block_count: (batch_size, nheads, cdiv(seqlen, BLOCK_M))
+        block_offset: (batch_size, nheads, cdiv(seqlen, BLOCK_M), NNZ_S)
+        column_count: (batch_size, nheads, cdiv(seqlen, BLOCK_M))
+        column_index: (batch_size, nheads, cdiv(seqlen, BLOCK_M), NNZ_V)
+        dropout_p: float. Dropout probability.
+        softmax_scale: float. The scaling of QK^T before applying softmax.
+            Default to 1 / sqrt(headdim).
+        causal: bool. Whether to apply causal attention mask (e.g., for auto-regressive modeling).
+        window_size: (left, right). If not (-1, -1), implements sliding window local attention.
```
Comment (LucasWilkinson) [inline vllm_flash_attn/flash_attn_interface.py:176]: is window still supported? this seems like a confusing overlap with the vertical slash mask but I could be mistaken

File: vllm_flash_attn/flash_attn_interface.py [hunk 6]
```diff
@@ -112,6 +136,78 @@ def _flash_attn_varlen_forward(
...
+    column_count and column_index for virtical sparsity patterns.
+
+    Arguments:
+        q: (batch_size, seqlen, nheads, headdim)
+        k: (batch_size, seqlen, nheads_k, headdim)
+        v: (batch_size, seqlen, nheads_k, headdim)
+        block_count: (batch_size, nheads, cdiv(seqlen, BLOCK_M))
+        block_offset: (batch_size, nheads, cdiv(seqlen, BLOCK_M), NNZ_S)
+        column_count: (batch_size, nheads, cdiv(seqlen, BLOCK_M))
+        column_index: (batch_size, nheads, cdiv(seqlen, BLOCK_M), NNZ_V)
```
Comment (LucasWilkinson) [inline vllm_flash_attn/flash_attn_interface.py:171]: would it be possible to have a diagram of the attention mask like in `flash_attn_func`, I think this would help users understand kernel better and thus make debugging easier

if there's a paper that describes these data structures (i.e. `block_count`, `block_offset`, `column_count`, `column_index`) that could be helpful too

**Review by LucasWilkinson — COMMENTED** [review]
Thanks for addressing my previous comments and thanks for the contribution! I did another pass and left some more comments. 

Overall the kernel seems quite cool but there seems to be alot of commented out code (thats appears to not be of the "un-comment for useful debug prints" style) that could use cleaning up. I think I caught most of it but another clean-up pass could be useful here.

File: vllm_flash_attn/flash_attn_interface.py (hunk 2 above)
Comment (LucasWilkinson) [reply vllm_flash_attn/flash_attn_interface.py:139]: I think a more verbose description of how the `block_count/block_offset` and `column_count/column_index` work could be really helpful here, will really help anyone trying to debug in the future.

File: csrc/flash_attn/flash_api.cpp (hunk 3 above)
Comment (LucasWilkinson) [reply csrc/flash_attn/flash_api.cpp:531]: Thanks!

File: vllm_flash_attn/flash_attn_interface.py (hunk 5 above)
Comment (minminsun) [reply vllm_flash_attn/flash_attn_interface.py:178]: No, sliding window is not supported in sparse_attn_func. I updated the docstring and added a TORCH_CHECK.

File: vllm_flash_attn/flash_attn_interface.py (hunk 6 above)
Comment (minminsun) [reply vllm_flash_attn/flash_attn_interface.py:173]: I found it's hard to explain the sparse related data structures, like block_count, block_offset, column_count, column_index, without some background introduction. I added the link of paper and code repo of MInference for reference in the docstring.

**Review by LucasWilkinson — COMMENTED** [review]
Thanks for all the changes! Overall looks pretty good to me, left a couple more (optional) nits.

My final concern is binary size (we are a bit sensitive to this in vLLM), do you know which head dims are actually being used (since theres only a limited set of models using this currently)? Ideally we'd only build and ship those for now

Can you get the DCO check to pass be signing off on the commits https://github.com/apps/dco. After that I think everything is good from my side, @WoosukKwon not sure if you want to take a look?

**Comment by minminsun** [comment]
> Thanks for all the changes! Overall looks pretty good to me, left a couple more (optional) nits.
> 
> My final concern is binary size (we are a bit sensitive to this in vLLM), do you know which head dims are actually being used (since theres only a limited set of models using this currently)? Ideally we'd only build and ship those for now
> 
> Can you get the DCO check to pass be signing off on the commits https://github.com/apps/dco. After that I think everything is good from my side, @WoosukKwon not sure if you want to take a look?

Keeping headdim 128 only is enough for us for now. 

The DCO check is now passed.  Thank you again for your thorough review and valuable suggestions!

**Review by LucasWilkinson — APPROVED** [review]
LGTM now, thanks!

**Comment by minminsun** [comment]
> LGTM now, thanks!

Thank you! Could you please help to get the PR merged? The vLLM PR https://github.com/vllm-project/vllm/pull/11844 depends on this.

**Comment by susu1210** [comment]
@minminsun  Hi, I was wonderring if the kernel supports multi-batch? 
```python
    batch_size = 2
    block_count = torch.tensor([NNZ_S] * batch_size * NUM_ROWS * num_heads, dtype=torch.int32).reshape(batch_size, num_heads, NUM_ROWS)
    column_count = torch.tensor([NNZ_V] * batch_size * NUM_ROWS * num_heads, dtype=torch.int32).reshape(batch_size, num_heads, NUM_ROWS)
    block_offset = torch.tensor([[i * block_size_N for i in range(NNZ_S)]] * batch_size * NUM_ROWS * num_heads, dtype=torch.int32).reshape(batch_size, num_heads, NUM_ROWS, NNZ_S)
    column_index = torch.tensor([[NNZ_S * block_size_N + i for i in range(NNZ_V)]] * batch_size * NUM_ROWS * num_heads, dtype=torch.int32).reshape(batch_size, num_heads, NUM_ROWS, NNZ_V)
```
block_offset column_index as mentioned above cannot pass the pytest examples. 
How do these variables work?

**Comment by LucasWilkinson** [comment]
@minminsun could you please expand the PR description to something like:

"Implements the kernel described in Appendix C.4.2 https://arxiv.org/abs/2407.02490 (as `sparse_attn_func`)"

Just so there will be a more useful commit message

**Comment by minminsun** [comment]
> @minminsun could you please expand the PR description to something like:
> 
> "Implements the kernel described in Appendix C.4.2 https://arxiv.org/abs/2407.02490 (as `sparse_attn_func`)"
> 
> Just so there will be a more useful commit message

Sure!

File: csrc/flash_attn/flash_api.cpp [hunk 7]
```diff
@@ -261,6 +338,195 @@ void set_params_alibi(Flash_fwd_params &params, const c10::optional<at::Tensor>
...
+    if (softcap > 0.f) { TORCH_CHECK(p_dropout == 0.f, "Softcapping does not support dropout for now"); }
+
+    // causal=true is the same as causal=false in this case
+    if (seqlen_q == 1 && !alibi_slopes_.has_value()) { is_causal = false; }
+
+    // Faster to transpose q from (b, 1, (nheads_kv ngroups), d) to (b, ngroups, nheads_kv, d) in this case
+    // H/t Daniel Haziza
+    const int seqlenq_ngroups_swapped = false;
+    const int ngroups = num_heads / num_heads_k;
+    if (seqlenq_ngroups_swapped) {
```
Comment (WoosukKwon) [inline csrc/flash_attn/flash_api.cpp:402]: QQ: If `seqlenq_ngroups_swapped` is always False, can we remove this part of code?

File: tests/test_vllm_flash_attn.py [hunk 8]
```diff
@@ -267,3 +267,63 @@ def test_varlen_with_paged_kv(
     )
     torch.testing.assert_close(output, ref_output, atol=2e-2, rtol=1e-2), \
         f"{torch.max(torch.abs(output - ref_output))}"
+
+@pytest.mark.parametrize("seq_lens", [(1023, 2049), (1023, 1023), (32, 32), (65, 65), (129, 129)])
```
Comment (WoosukKwon) [inline tests/test_vllm_flash_attn.py:271]: Thanks for adding the unit test! Can you please also include query_len=1 cases like (1, 1) and (1, 1024)? FA includes some special code paths for query_len=1, so it would be good to check here.

File: csrc/flash_attn/flash_api.cpp [hunk 9]
```diff
@@ -261,6 +338,195 @@ void set_params_alibi(Flash_fwd_params &params, const c10::optional<at::Tensor>
 #endif
 }
 
+std::vector<at::Tensor>
+mha_fwd_sparse(at::Tensor &q,         // batch_size x seqlen_q x num_heads x head_size
```
Comment (WoosukKwon) [inline csrc/flash_attn/flash_api.cpp:342]: Would it be possible to add the varlen version of this kernel? I think that will help simplify the vLLM integration a lot.

File: csrc/flash_attn/flash_api.cpp (hunk 7 above)
Comment (minminsun) [reply csrc/flash_attn/flash_api.cpp]: OK, removed.

File: tests/test_vllm_flash_attn.py (hunk 8 above)
Comment (minminsun) [reply tests/test_vllm_flash_attn.py]: Sure, added.

File: csrc/flash_attn/flash_api.cpp (hunk 9 above)
Comment (minminsun) [reply csrc/flash_attn/flash_api.cpp:342]: Sure, interface sparse_attn_varlen_func has been added.

File: csrc/flash_attn/flash_api.cpp [hunk 10]
```diff
@@ -261,6 +338,394 @@ void set_params_alibi(Flash_fwd_params &params, const c10::optional<at::Tensor>
...
+    //                  seqused_k.has_value() ? seqused_k.value().data_ptr() : nullptr,
+    //                  return_softmax ? p.data_ptr() : nullptr,
+    //                  softmax_lse.data_ptr(),
+    //                  p_dropout,
+    //                  softmax_scale,
+    //                  -1,
+    //                  -1,
+    //                  softcap,
+    //                  seqlenq_ngroups_swapped,
+    //                  /*unpadded_lse*/true);
```
Comment (LucasWilkinson) [inline csrc/flash_attn/flash_api.cpp:666]: nit: remove cruft (commented out `set_params_fprop(params,`)

**Review by LucasWilkinson — COMMENTED** [review]
Thanks for adding the varlen api! it appears the `minf.py` is not used? am I missing something or can this be removed?

File: csrc/flash_attn/flash_api.cpp (hunk 10 above)
Comment (minminsun) [reply csrc/flash_attn/flash_api.cpp]: Thanks, removed.

**Comment by minminsun** [comment]
> Thanks for adding the varlen api! it appears the `minf.py` is not used? am I missing something or can this be removed?

No, it's not used. Removed.

**Comment by yzh119** [comment]
Hi @minminsun @LucasWilkinson @WoosukKwon , I just saw this issue, you might be interested in flashinfer's sparse attention implementation (which supports fine-grained block size, and both fa2/3 templates), you can try this feature using the sparse attention API in flashinfer: https://docs.flashinfer.ai/api/sparse.html. and it was used in projects such as [quest](https://arxiv.org/pdf/2405.19846).

**Comment by LucasWilkinson** [comment]
Cool thanks for the letting us know, ill check it out!

**Comment by GWS0428** [comment]
Hi, I recently tried using the sparse kernel, but I'm encountering mismatch errors in many cases. For example:

```    
test_sparse_attention_varlen(
        seq_lens=[(512, 1025)],
        head_size=128,
        dtype=torch.float16,
    )
    
    Mismatched elements: 8656 / 65536 (13.2%)
Greatest absolute difference: nan at index (0, 0, 0) (up to 0.02 allowed)
Greatest relative difference: nan at index (0, 0, 0) (up to 0.01 allowed)
```

When I inspect the out variable, it contains a significant number of NaN values. Adding the following code at the end of the test confirms this:

```    # nan check
    if torch.isnan(out).any():
        print("NaN test : Custom kernel output contains NaN values.")
    else:
        print("NaN test : Custom kernel output does not contain NaN values.")
```

While debugging the kernel, I noticed that acc_s appears to be corrupted during non-masked steps, showing irregular shapes like:

```
---- acc_s / tensor data (Thread 0) ----
Logical shape of tensor view: ((-629100032, 1), (2, 16))
```

Could you verify the correctness of the kernel implementation? If you're not actively maintaining it, could you guide me on how to fix or debug this further?

Thanks!

**Comment by zongy17** [comment]
Hi @minminsun, can MInference sparse attention be enabled in vLLM versions with v1 scheduling (i.e., no separation of prefill and decode phases)? If it could, what environment variable can be set to enable it? Thanks!
